citrix_hypervisor/datadog_checks/citrix_hypervisor/check.py

```diff
@@ -0,0 +1,182 @@
+        # type: (Dict) -> None
+        legends = data['meta']['legend']
+        values = data['data'][0]['values']
+        for i in range(len(legends)):
+            metric_name, tags = build_metric(legends[i], self.log)
+
+            if metric_name is not None:
+                self.gauge(metric_name, values[i], tags=self.tags + self._additional_tags + tags)
+
+    def _session_login(self, server):
+        # type: (Any) -> Dict[str, str]
+        try:
+            session = server.session.login_with_password(
+                self.instance.get('username', ''), self.instance.get('password', '')
+            )
+            return session
+        except Exception as e:
+            self.log.warning(e)
+        return {}
+
+    def _get_master_session(self, session):
+        # type: (Dict[str, str]) -> Dict[str, str]
+        # {'Status': 'Failure', 'ErrorDescription': ['HOST_IS_SLAVE', '192.168.101.102']}
+        master_address = session['ErrorDescription'][1]
+        if not master_address.startswith('http://'):
+            master_address = 'http://' + master_address
+        master_xenserver = xmlrpclib.Server(master_address)
+
+        # Master credentials can be different, we can specify new `master_username` and
+        # `master_password` options later if requested
+        master_session = self._session_login(master_xenserver)
+
+        if master_session.get('Status') == 'Success':
+            self.xenserver = master_xenserver
+            return master_session
+
+        return {}
+
+    def open_session(self):
+        # type: () -> Dict[str, str]
+        try:
+            self.xenserver = xmlrpclib.Server(self._base_url)
+        except Exception as e:
+            self.log.warning(str(e))
+            return {}
+
+        # See reference https://xapi-project.github.io/xen-api/classes/session.html
+        session = self._session_login(self.xenserver)
+
+        if session.get('Status') == 'Failure':
+            if 'SESSION_AUTHENTICATION_FAILED' in session.get('ErrorDescription', []):
+                self.log.warning('Connection failed with xmlrpc: %s', str(session['ErrorDescription']))
+                return {}
+            if 'HOST_IS_SLAVE' in session.get('ErrorDescription', []):
+                # If xenserver host is a slave, open a connection with the pool master
+                # {'Status': 'Failure', 'ErrorDescription': ['HOST_IS_SLAVE', '192.168.101.102']}
+                self.log.debug('Host is a slave, trying to connect to the pool master')
+                self._additional_tags.append('server_type:slave')
+                return self._get_master_session(session)
+
+            self.log.warning('Unknown xmlrpc error: %s', str(session))
+            return {}
+
+        self._additional_tags.append('server_type:master')
+
+        return session
+
+    def check(self, _):
+        # type: (Any) -> None
+        self._additional_tags = []
+        session = self.open_session()
```

Comment:
if `self.open_session()` returns {} on exception, should the check continue if the session wasn't successfully opened?

Reply:
Yes, the check should continue running because if the instance is configured for a pool slave host, and we can't connect to the pool master (bad master credentials for example), we can't open a session. However, the HTTP endpoint is still available for the pool slave, so we can still get all the metrics.

The session opening is currently only used to get a tag, and can be used to get other info later